manager/scheduler/scheduler.go

```diff
@@ -427,30 +446,97 @@ func (s *Scheduler) taskFitNode(ctx context.Context, t *api.Task, nodeID string)
+	// tasks in the group are equal in terms of the fields the constraint
+	// filters consider.
+	var t *api.Task
+	for _, t = range taskGroup {
+		break
+	}
+
 	s.pipeline.SetTask(t)
-	n, _ := s.nodeHeap.findMin(s.pipeline.Process, s.scanAllNodes)
-	if n == nil {
-		log.G(ctx).WithField("task.id", t.ID).Debug("No suitable node available for task")
-		return nil
+
+	nodeLess := func(a *NodeInfo, b *NodeInfo) bool {
+		sameServiceTasksOnA := 0
+		sameServiceTasksOnB := 0
+
+		// TODO(aaronl): NodeInfo should keep a map of serviceID->[]task,
+		// so that checking how many tasks from a given service are
+		// running on the node is cheap.
+		for _, aTask := range a.Tasks {
+			if aTask.ServiceID == t.ServiceID {
+				sameServiceTasksOnA++
+			}
+		}
+		for _, bTask := range b.Tasks {
+			if bTask.ServiceID == t.ServiceID {
+				sameServiceTasksOnB++
+			}
+		}
+
+		if sameServiceTasksOnA < sameServiceTasksOnB {
+			return true
+		}
+		if sameServiceTasksOnB > sameServiceTasksOnA {
+			return false
+		}
+
+		// Total number of tasks breaks ties.
+		return len(a.Tasks) < len(b.Tasks)
 	}
 
-	log.G(ctx).WithField("task.id", t.ID).Debugf("Assigning to node %s", n.ID)
-	newT := *t
-	newT.NodeID = n.ID
-	newT.Status = api.TaskStatus{
-		State:     api.TaskStateAssigned,
-		Timestamp: ptypes.MustTimestampProto(time.Now()),
-		Message:   "scheduler assigned task to node",
+	nodes := s.nodeHeap.findBestNodes(len(taskGroup), s.pipeline.Process, nodeLess)
+	if len(nodes) == 0 {
+		for _, t := range taskGroup {
+			log.G(ctx).WithField("task.id", t.ID).Debug("no suitable node available for task")
+			s.enqueue(t)
+		}
+		return
 	}
-	s.allTasks[t.ID] = &newT
 
-	nodeInfo, err := s.nodeHeap.nodeInfo(n.ID)
-	if err == nil && nodeInfo.addTask(&newT) {
-		s.nodeHeap.updateNode(nodeInfo)
+	nodeIter := 0
+	for taskID, t := range taskGroup {
+		if nodeIter > len(nodes) {
+			origNodeIter := nodeIter
+			// TODO(aaronl): Mark nodes that no longer meet the
+			// constraints to avoid repeated evaluations.
+			for !s.pipeline.Process(nodes[nodeIter%len(nodes)]) {
+				nodeIter++
+				if nodeIter-origNodeIter == len(nodes) {
+					// None of the nodes meet the constraints anymore.
+					for _, t := range taskGroup {
+						log.G(ctx).WithField("task.id", t.ID).Debug("no suitable node available for task")
+						s.enqueue(t)
+					}
+					return
+				}
+			}
+		}
+
+		n := nodes[nodeIter%len(nodes)]
+
+		log.G(ctx).WithField("task.id", t.ID).Debugf("assigning to node %s", n.ID)
+		newT := *t
+		newT.NodeID = n.ID
+		newT.Status = api.TaskStatus{
+			State:     api.TaskStateAssigned,
+			Timestamp: ptypes.MustTimestampProto(time.Now()),
+			Message:   "scheduler assigned task to node",
+		}
+		s.allTasks[t.ID] = &newT
+
+		nodeInfo, err := s.nodeHeap.nodeInfo(n.ID)
+		if err == nil && nodeInfo.addTask(&newT) {
+			s.nodeHeap.updateNode(nodeInfo)
+		}
+
+		schedulingDecisions[taskID] = schedulingDecision{old: t, new: &newT}
+		delete(taskGroup, taskID)
```

Comment:
Also fixed. The reason the tests didn't catch it is that they set up a situation where the assignments are supposed to go to particular node, because that node is more lightly loaded.